packages/uix-core/src/tunnel/tunnel-messenger.ts

```diff
@@ -0,0 +1,99 @@
+import { NS_ROOT, VERSION } from "../constants";
+import { isPlainObject } from "../value-assertions";
+import { WrappedMessage, isWrapped, wrap, unwrap } from "../message-wrapper";
+import { HandshakeAcceptedTicket, HandshakeOfferedTicket } from "../tickets";
+
+type Handshake = HandshakeAcceptedTicket | HandshakeOfferedTicket;
+type HandshakeAccepted = WrappedMessage<HandshakeAcceptedTicket>;
+type HandshakeOffered = WrappedMessage<HandshakeOfferedTicket>;
+type HandshakeMessage = HandshakeAccepted | HandshakeOffered;
+
+export class TunnelMessenger {
+  private myOrigin: string;
+  private remoteOrigin: string;
+  private logger: Console;
+  private versionWarnings = new Set<string>();
```

Comment:
Do we really need a state to manage warnings?

Reply:
With multiple extensions running and connecting additional frames, we may get repetitive version mismatch warnings. The state here just ensures only one warning for each mismatch.